subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -441,8 +414,8 @@ static struct tls_context *tls_alloc(void)
 		mbedtls_pk_init(&tls->priv_key);
 #endif
 
-#if defined(MBEDTLS_DEBUG_C) && (CONFIG_NET_SOCKETS_LOG_LEVEL >= LOG_LEVEL_DBG)
-		mbedtls_ssl_conf_dbg(&tls->config, tls_debug, NULL);
+#if defined(CONFIG_MBEDTLS_DEBUG)
```

Comment:
This change prevents non-builtin MBed-TLS users from benefitting from these logging improvements.

If you changed it to:
`#if defined(MBEDTLS_DEBUG_C)`
then it will still work for both cases.

Reply:
Any comments @mniestroj ?

Reply:
> This change prevents non-builtin MBed-TLS users from benefitting from these logging improvements.

Yes, you are right. Initially I was trying to keep compatibility with "non-builtin" (`CONFIG_MBEDTLS_LIBRARY`). However the more I looked into that, the more existing issues I have found with that variant:
1. `modules/mbedtls/CMakeLists.txt` compiles in `zephyr_init.c` only for `CONFIG_MBEDTLS_BUILTIN=y`, which means that heap allocation is configured only for builtin variant,
2. `subsys/net/lib/sockets/sockets_tls.c` enables log messages with following code:
```
#if defined(MBEDTLS_DEBUG_C) && (CONFIG_NET_SOCKETS_LOG_LEVEL >= LOG_LEVEL_DBG)
	mbedtls_debug_set_threshold(CONFIG_MBEDTLS_DEBUG_LEVEL);
#endif
```
which in theory is enabled for external (non-builtin) mbedTLS variant, but if you look closer `CONFIG_MBEDTLS_DEBUG_LEVEL` is defined only when `CONFIG_MBEDTLS_DEBUG` is defined, which is only when `CONFIG_MBEDTLS_BUILTIN` is selected. So with `CONFIG_MBEDTLS_LIBRARY=y` Zephyr TLS socket glue layer wouldn't compile IMO (not sure, didn't try).
3. Any commits strictly related to `CONFIG_MBEDTLS_LIBRARY=y` are from few years back, so there is probably little or no users nowadays, which might be the reason why Zephyr TLS sockets would no longer build with external mbedTLS (see point 2.).
4. I am not aware of any build-test (or even a configuration outside of twister) of external mbedTLS library with Zephyr, so trying to setup this locally requires some more effort.

>If you changed it to:
>```
>#if defined(MBEDTLS_DEBUG_C)
>```
>then it will still work for both cases.

Unfortunately this is not enough. `CMakeLists.txt` needs to be updated to build "improved" `zephyr_mbedtls_debug()` function. As both: 1) heap setup and 2) setting mbedTLS log level, are not properly supported with `CONFIG_MBEDTLS_LIBRARY` (external mbedTLS), then I decided not to **try** to make it work.

@plskeggs Are you using external mbedTLS library actively with Zephyr?

Reply:
Yes, Nordic's nRF Connect SDK has its own fork of Mbed-TLS, with some helper code that takes care of the equivalent of `zephyr_init.c`.  I have a pending PR to that which adds calling `mbedtls_debug_set_threshold()` as needed.

I agree `CMakeLists.txt` would also need to change to compile in your new `zephyr_mbedtls_debug()` function.

Reply:
I have a feeling that you are not using neither `CONFIG_MBEDTLS_BUILTIN` nor `CONFIG_MBEDTLS_LIBRARY`, but all the config symbols come from another Kconfig (not Zephyr upstream). Is that right? In that case I have really limited room for improvement, because the problem is not related to Zephyr upstream.

Reply:
No, we use `CONFIG_MBEDTLS_LIBRARY`.

Reply:
If you'd rather not accommodate this use case, I could put in a PR after yours merges to address my issues.

Reply:
> No, we use CONFIG_MBEDTLS_LIBRARY.

Interesting. Could you provide me with example application that uses it, so I can give it a quick try?

>If you'd rather not accommodate this use case, I could put in a PR after yours merges to address my issues.

If you're okay with that, I could do my best to review such PR.

Reply:
Sorry, my mistake.  It's not `CONFIG_MBEDTLS_LIBRARY` that we use, it's `CONFIG_MBEDTLS_TLS_LIBRARY` which is defined in one of our external libraries.

So, I'll wait for your PR to merge before I proceed with my own.

Reply:
@tejlmand -- yes, that does work -- I tested it today with the commits from this PR.  However, without `CMakeLists.txt` changes, the new `debug.c` here does not get built.  And without `sockets_tls.c` changes, because we don't need `zephyr_init.c`, the call to `mbedtls_debug_set_threshold()` does not get made either.

I think on our side we should discuss either switching to CONFIG_MBEDTLS_LIBRARY, or define a new MBEDTLS_IMPLEMENTATION to cover our case, which would then build `debug.c` and call `mbedtls_debug_set_threshold()`.